py-polars/polars/internals/dataframe/frame.py

```diff
@@ -2451,50 +2451,61 @@ def describe(self: DF) -> DF:
         ... )
         >>> df.describe()
-        shape: (5, 4)
-        ┌──────────┬──────────┬─────┬──────┐
-        │ describe ┆ a        ┆ b   ┆ c    │
-        │ ---      ┆ ---      ┆ --- ┆ ---  │
-        │ str      ┆ f64      ┆ f64 ┆ f64  │
-        ╞══════════╪══════════╪═════╪══════╡
-        │ mean     ┆ 2.266667 ┆ 5.0 ┆ null │
-        ├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌┤
-        │ std      ┆ 1.101514 ┆ 1.0 ┆ null │
-        ├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌┤
-        │ min      ┆ 1.0      ┆ 4.0 ┆ 0.0  │
-        ├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌┤
-        │ max      ┆ 3.0      ┆ 6.0 ┆ 1.0  │
-        ├╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌┤
-        │ median   ┆ 2.8      ┆ 5.0 ┆ null │
-        └──────────┴──────────┴─────┴──────┘
-
-        """
-
-        def describe_cast(self: DF) -> DF:
+        shape: (7, 7)
+        ┌────────────┬──────────┬──────────┬──────┬──────┬──────┬────────────┐
+        │ describe   ┆ a        ┆ b        ┆ c    ┆ d    ┆ e    ┆ f          │
+        │ ---        ┆ ---      ┆ ---      ┆ ---  ┆ ---  ┆ ---  ┆ ---        │
+        │ str        ┆ f64      ┆ f64      ┆ f64  ┆ str  ┆ str  ┆ str        │
+        ╞════════════╪══════════╪══════════╪══════╪══════╪══════╪════════════╡
+        │ count      ┆ 3.0      ┆ 3.0      ┆ 3.0  ┆ 3    ┆ 3    ┆ 3          │
+        ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ null_count ┆ 0.0      ┆ 1.0      ┆ 0.0  ┆ 1    ┆ 1    ┆ 0          │
+        ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ mean       ┆ 2.266667 ┆ 4.5      ┆ null ┆ null ┆ null ┆ null       │
+        ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ std        ┆ 1.101514 ┆ 0.707107 ┆ null ┆ null ┆ null ┆ null       │
+        ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ min        ┆ 1.0      ┆ 4.0      ┆ 0.0  ┆ b    ┆ eur  ┆ 2020-01-01 │
+        ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ max        ┆ 3.0      ┆ 5.0      ┆ 1.0  ┆ c    ┆ usd  ┆ 2022-01-01 │
+        ├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
+        │ median     ┆ 2.8      ┆ 4.5      ┆ null ┆ null ┆ null ┆ null       │
+        └────────────┴──────────┴──────────┴──────┴──────┴──────┴────────────┘
+
+        """
+
+        def describe_cast(stat: DF) -> DF:
             columns = []
-            for s in self:
-                if s.is_numeric() or s.is_boolean():
-                    columns.append(s.cast(float))
+            for i, s in enumerate(self.columns):
+                if self[s].is_numeric() or self[s].is_boolean():
+                    columns.append(stat[:, i].cast(float))
                 else:
-                    columns.append(s)
+                    # for dates, strings, etc, we cast to string so that all
+                    # statistics can be shown
+                    columns.append(stat[:, i].cast(str))
```

Comment:
Was `null_count` not available for `dates? That is a bug then.

Reply:
It is available, see the doc string example here: https://github.com/pola-rs/polars/pull/4820/files#diff-0a0c19eef131f26f17dab6b0f66bb23abd81cb696e646626e2045794e274f6beR2476

It is just that because we report each dataframe column as a column in describe, that we get a conflict for the column types when having both things like count and null_count, which are integers, with min and max, which are dependent on the column type. So that is why casting to string happens.

Reply:
Oh, right. And were talking about very small series here. I understand.